File: python/app/services/template_matcher.py
from __future__ import annotations

import logging
from pathlib import Path
from typing import Dict, List, Optional, Tuple

import cv2
import numpy as np
from app.models.schemas import MatchItem, BBox, Point
from app.services.template_store import TemplateStore

logger = logging.getLogger(__name__)


class TemplateMatcher:
    def __init__(self, store: TemplateStore):
        self._store = store
        self._cache: Dict[str, np.ndarray] = {}
        self._load_images()

    def _load_images(self) -> None:
        self._cache.clear()
        for info in self._store.list_all():
            img_path = self._store.get_image_path(info.name)
            if img_path is not None:
                img = cv2.imread(str(img_path), cv2.IMREAD_COLOR)
                if img is not None:
                    self._cache[info.name] = img
                else:
                    logger.warning("Failed to load template image: %s", img_path)
        logger.info("Loaded %d template images into cache", len(self._cache))

    def reload(self) -> None:
        self._store.reload()
        self._load_images()

    def match_all(self, image: np.ndarray) -> List[MatchItem]:
        results: List[MatchItem] = []
        for info in self._store.list_all():
            template_img = self._cache.get(info.name)
            if template_img is None:
                continue
            matches = self._match_template(image, template_img, info.name, info.threshold)
            results.extend(matches)
        return results

    def match(self, image: np.ndarray, template_names: List[str]) -> List[MatchItem]:
        results: List[MatchItem] = []
        for name in template_names:
            info = self._store.get(name)
            if info is None:
                continue
            template_img = self._cache.get(name)
            if template_img is None:
                continue
            matches = self._match_template(image, template_img, name, info.threshold)
            results.extend(matches)
        return results

    def _match_template(
        self,
        image: np.ndarray,
        template: np.ndarray,
        name: str,
        threshold: float,
    ) -> List[MatchItem]:
        th, tw = template.shape[:2]
        ih, iw = image.shape[:2]

        all_boxes: List[Tuple[int, int, int, int, float]] = []
        scales = [0.85, 0.9, 0.95, 1.0, 1.05, 1.1, 1.15]

        for scale in scales:
            sw = int(tw * scale)
            sh = int(th * scale)
            if sw <= 0 or sh <= 0 or sw > iw or sh > ih:
                continue

            if scale != 1.0:
                scaled = cv2.resize(template, (sw, sh), interpolation=cv2.INTER_LINEAR)
            else:
                scaled = template

            result = cv2.matchTemplate(image, scaled, cv2.TM_CCOEFF_NORMED)
            locations = np.where(result >= threshold)

            for pt_y, pt_x in zip(*locations):
                conf = float(result[pt_y, pt_x])
                all_boxes.append((int(pt_x), int(pt_y), sw, sh, conf))

        filtered = self._nms(all_boxes, iou_threshold=0.3)

        items: List[MatchItem] = []
        for (bx, by, bw, bh, conf) in filtered:
            items.append(MatchItem(
                template=name,
                confidence=round(conf, 4),
                bbox=BBox(x=bx, y=by, width=bw, height=bh),
                center=Point(x=bx + bw // 2, y=by + bh // 2),
            ))
        return items

    @staticmethod
    def _nms(
        boxes: List[Tuple[int, int, int, int, float]],
        iou_threshold: float = 0.3,
    ) -> List[Tuple[int, int, int, int, float]]:
        if not boxes:
            return []

        sorted_boxes = sorted(boxes, key=lambda b: b[4], reverse=True)
        keep: List[Tuple[int, int, int, int, float]] = []

        while sorted_boxes:
            best = sorted_boxes.pop(0)
            keep.append(best)
            remaining: List[Tuple[int, int, int, int, float]] = []
            for box in sorted_boxes:
                if _iou(best, box) < iou_threshold:
                    remaining.append(box)
            sorted_boxes = remaining

        return keep


def _iou(
    a: Tuple[int, int, int, int, float],
    b: Tuple[int, int, int, int, float],
) -> float:
    ax1, ay1, aw, ah, _ = a
    bx1, by1, bw, bh, _ = b
    ax2, ay2 = ax1 + aw, ay1 + ah
    bx2, by2 = bx1 + bw, by1 + bh

    ix1 = max(ax1, bx1)
    iy1 = max(ay1, by1)
    ix2 = min(ax2, bx2)
    iy2 = min(ay2, by2)

    if ix2 <= ix1 or iy2 <= iy1:
        return 0.0

    inter = (ix2 - ix1) * (iy2 - iy1)
    union = aw * ah + bw * bh - inter
    return inter / union if union > 0 else 0.0

File: src/SGMDTXTools.Core/Services/TemplateMatcher.cs
using OpenCvSharp;
using Serilog;
using SGMDTXTools.Core.Models;

namespace SGMDTXTools.Core.Services;

public class TemplateMatcher : IDisposable
{
    private readonly ILogger _log;
    private readonly TemplateStore _store;
    private readonly Dictionary<string, Mat> _cache = new();
    private bool _disposed;

    private static readonly double[] Scales = [0.85, 0.9, 0.95, 1.0, 1.05, 1.1, 1.15];

    public TemplateMatcher(ILogger logger, TemplateStore store)
    {
        _log = logger.ForContext<TemplateMatcher>();
        _store = store;
        LoadImages();
    }

    private void LoadImages()
    {
        DisposeCache();

        foreach (var info in _store.ListAll())
        {
            var imgPath = _store.GetImagePath(info.Name);
            if (imgPath == null) continue;

            var img = Cv2.ImRead(imgPath, ImreadModes.Color);
            if (img.Empty())
            {
                _log.Warning("无法加载模板图片: {Path}", imgPath);
                img.Dispose();
                continue;
            }
            _cache[info.Name] = img;
        }

        _log.Information("加载了 {Count} 个模板图片到缓存", _cache.Count);
    }

    public void Reload()
    {
        _store.Reload();
        LoadImages();
    }

    public List<MatchItem> MatchAll(Mat image)
    {
        var results = new List<MatchItem>();
        foreach (var info in _store.ListAll())
        {
            if (!_cache.TryGetValue(info.Name, out var tmpl)) continue;
            results.AddRange(MatchTemplate(image, tmpl, info.Name, info.Threshold));
        }
        return results;
    }

    public List<MatchItem> Match(Mat image, string[] templateNames)
    {
        var results = new List<MatchItem>();
        foreach (var name in templateNames)
        {
            var info = _store.Get(name);
            if (info == null) continue;
            if (!_cache.TryGetValue(name, out var tmpl)) continue;
            results.AddRange(MatchTemplate(image, tmpl, name, info.Threshold));
        }
        return results;
    }

    private List<MatchItem> MatchTemplate(Mat image, Mat template, string name, double threshold)
    {
        int th = template.Rows, tw = template.Cols;
        int ih = image.Rows, iw = image.Cols;

        var allBoxes = new List<(int X, int Y, int W, int H, double Conf)>();

        foreach (var scale in Scales)
        {
            int sw = (int)(tw * scale);
            int sh = (int)(th * scale);
            if (sw <= 0 || sh <= 0 || sw > iw || sh > ih) continue;

            Mat scaled;
            if (Math.Abs(scale - 1.0) < 0.001)
            {
                scaled = template;
            }
            else
            {
                scaled = new Mat();
                Cv2.Resize(template, scaled, new Size(sw, sh), interpolation: InterpolationFlags.Linear);
            }

            using var result = new Mat();
            Cv2.MatchTemplate(image, scaled, result, TemplateMatchModes.CCoeffNormed);

            // 找到所有超过阈值的位置
            for (int row = 0; row < result.Rows; row++)
            {
                for (int col = 0; col < result.Cols; col++)
                {
                    float conf = result.At<float>(row, col);
                    if (conf >= threshold)
                    {
                        allBoxes.Add((col, row, sw, sh, conf));
                    }
                }
            }

            if (Math.Abs(scale - 1.0) >= 0.001)
                scaled.Dispose();
        }

        var filtered = Nms(allBoxes, 0.3);

        var items = new List<MatchItem>();
        foreach (var (bx, by, bw, bh, conf) in filtered)
        {
            items.Add(new MatchItem
            {
                Template = name,
                Confidence = Math.Round(conf, 4),
                Bbox = new BoundingBox { X = bx, Y = by, Width = bw, Height = bh },
                Center = new PointCoord { X = bx + bw / 2, Y = by + bh / 2 }
            });
        }
        return items;
    }

    private static List<(int X, int Y, int W, int H, double Conf)> Nms(
        List<(int X, int Y, int W, int H, double Conf)> boxes, double iouThreshold)
    {
        if (boxes.Count == 0) return [];

        var sorted = boxes.OrderByDescending(b => b.Conf).ToList();
        var keep = new List<(int X, int Y, int W, int H, double Conf)>();

        while (sorted.Count > 0)
        {
            var best = sorted[0];
            sorted.RemoveAt(0);
            keep.Add(best);

            sorted.RemoveAll(box => Iou(best, box) >= iouThreshold);
        }

        return keep;
    }

    private static double Iou(
        (int X, int Y, int W, int H, double Conf) a,
        (int X, int Y, int W, int H, double Conf) b)
    {
        int ax2 = a.X + a.W, ay2 = a.Y + a.H;
        int bx2 = b.X + b.W, by2 = b.Y + b.H;

        int ix1 = Math.Max(a.X, b.X);
        int iy1 = Math.Max(a.Y, b.Y);
        int ix2 = Math.Min(ax2, bx2);
        int iy2 = Math.Min(ay2, by2);

        if (ix2 <= ix1 || iy2 <= iy1) return 0.0;

        double inter = (ix2 - ix1) * (iy2 - iy1);
        double union = (double)a.W * a.H + (double)b.W * b.H - inter;
        return union > 0 ? inter / union : 0.0;
    }

    private void DisposeCache()
    {
        foreach (var mat in _cache.Values)
            mat.Dispose();
        _cache.Clear();
    }

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        DisposeCache();
    }
}
